File: android-app/core/src/main/kotlin/com/evangeline/reader/narration/ReadingOrder.kt
package com.evangeline.reader.narration

import com.evangeline.reader.model.COLUMN_GUTTER_MIN
import com.evangeline.reader.model.DocumentBlock
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

private const val MIN_BLOCKS_FOR_COLUMNS = 4
private const val MIN_BLOCKS_PER_COLUMN = 2
private const val GUTTER_MIN_X = 0.28
private const val GUTTER_MAX_X = 0.72
private const val MIN_COLUMN_OVERLAP = 0.45
private const val SAME_LINE_RATIO = 0.4

private fun DocumentBlock.centreX(): Double = bounds.x + bounds.width / 2
private fun DocumentBlock.centreY(): Double = bounds.y + bounds.height / 2

/** Top to bottom, then left to right for blocks sharing a line. */
private fun sortVisual(blocks: List<DocumentBlock>): List<DocumentBlock> =
    blocks.sortedWith { a, b ->
        val dy = a.centreY() - b.centreY()
        val sameLine = abs(dy) <= min(a.bounds.height, b.bounds.height) * SAME_LINE_RATIO
        if (sameLine) a.bounds.x.compareTo(b.bounds.x) else dy.compareTo(0.0)
    }

private data class Span(val min: Double, val max: Double)

private fun verticalSpan(blocks: List<DocumentBlock>): Span {
    var lowest = 1.0
    var highest = 0.0
    for (block in blocks) {
        lowest = min(lowest, block.bounds.y)
        highest = max(highest, block.bounds.y + block.bounds.height)
    }
    return Span(lowest, highest)
}

/** The midpoint of the widest horizontal gap between block centres. */
private fun widestGapCentre(blocks: List<DocumentBlock>): Pair<Double, Double> {
    val centres = blocks.map { it.centreX() }.sorted()
    var bestGap = 0.0
    var splitAt = -1.0
    for (i in 1 until centres.size) {
        val gap = centres[i] - centres[i - 1]
        if (gap > bestGap) {
            bestGap = gap
            splitAt = (centres[i - 1] + centres[i]) / 2
        }
    }
    return bestGap to splitAt
}

/**
 * Two column runs only count as columns when they sit side by side over roughly
 * the same vertical range. Without this a heading above a block of text would
 * split into phantom columns and be read out of order.
 */
private fun runsSideBySide(left: List<DocumentBlock>, right: List<DocumentBlock>): Boolean {
    val leftSpan = verticalSpan(left)
    val rightSpan = verticalSpan(right)
    val overlap = min(leftSpan.max, rightSpan.max) - max(leftSpan.min, rightSpan.min)
    val union = max(leftSpan.max, rightSpan.max) - min(leftSpan.min, rightSpan.min)
    return union > 0 && overlap / union >= MIN_COLUMN_OVERLAP
}

private fun splitByGutter(blocks: List<DocumentBlock>): List<List<DocumentBlock>> {
    if (blocks.size < MIN_BLOCKS_FOR_COLUMNS) return listOf(blocks)

    val (bestGap, splitAt) = widestGapCentre(blocks)
    if (bestGap < COLUMN_GUTTER_MIN || splitAt < GUTTER_MIN_X || splitAt > GUTTER_MAX_X) {
        return listOf(blocks)
    }

    val (left, right) = blocks.partition { it.centreX() < splitAt }
    if (left.size < MIN_BLOCKS_PER_COLUMN || right.size < MIN_BLOCKS_PER_COLUMN) {
        return listOf(blocks)
    }
    return if (runsSideBySide(left, right)) listOf(left, right) else listOf(blocks)
}

/**
 * Puts a page's text blocks into the order a person would read them: each
 * column finished before the next begins (FR-005).
 *
 * Blocks carrying no visible text are dropped so they never become silent
 * passages.
 */
fun orderBlocks(blocks: List<DocumentBlock>): List<DocumentBlock> {
    val usable = blocks.filter { it.text.isNotBlank() }
    return splitByGutter(usable).flatMap(::sortVisual)
}

File: reader-app/src/reader/narration/reading-order.ts
import { COLUMN_GUTTER_MIN } from "../core/config.ts";
import type { DocumentBlock } from "../core/types.ts";

function centerX(block: DocumentBlock): number {
  return block.bounds.x + block.bounds.width / 2;
}

function centerY(block: DocumentBlock): number {
  return block.bounds.y + block.bounds.height / 2;
}

function sortVisual(blocks: DocumentBlock[]): DocumentBlock[] {
  return [...blocks].sort((a, b) => {
    const dy = centerY(a) - centerY(b);
    if (Math.abs(dy) > Math.min(a.bounds.height, b.bounds.height) * 0.4) {
      return dy;
    }
    return a.bounds.x - b.bounds.x;
  });
}

function splitByGutter(blocks: DocumentBlock[]): DocumentBlock[][] {
  if (blocks.length < 4) return [blocks];

  const centers = blocks.map(centerX).sort((a, b) => a - b);
  let bestGap = 0;
  let splitAt = -1;
  for (let i = 1; i < centers.length; i++) {
    const left = centers[i - 1] ?? 0;
    const right = centers[i] ?? 0;
    const gap = right - left;
    if (gap > bestGap) {
      bestGap = gap;
      splitAt = (left + right) / 2;
    }
  }

  if (bestGap < COLUMN_GUTTER_MIN || splitAt < 0.28 || splitAt > 0.72) {
    return [blocks];
  }

  const left: DocumentBlock[] = [];
  const right: DocumentBlock[] = [];
  for (const block of blocks) {
    if (centerX(block) < splitAt) left.push(block);
    else right.push(block);
  }

  if (left.length < 2 || right.length < 2) return [blocks];

  const leftSpan = verticalSpan(left);
  const rightSpan = verticalSpan(right);
  const overlap =
    Math.min(leftSpan.max, rightSpan.max) - Math.max(leftSpan.min, rightSpan.min);
  const union =
    Math.max(leftSpan.max, rightSpan.max) - Math.min(leftSpan.min, rightSpan.min);
  if (union <= 0 || overlap / union < 0.45) return [blocks];

  return [left, right];
}

function verticalSpan(blocks: DocumentBlock[]): { min: number; max: number } {
  let min = 1;
  let max = 0;
  for (const block of blocks) {
    min = Math.min(min, block.bounds.y);
    max = Math.max(max, block.bounds.y + block.bounds.height);
  }
  return { min, max };
}

export function orderBlocks(blocks: DocumentBlock[]): DocumentBlock[] {
  const usable = blocks.filter((block) => block.text.trim().length > 0);
  const columns = splitByGutter(usable);
  return columns.flatMap(sortVisual);
}

// A table's own column gap can look exactly like a two-column page's gutter
// to a purely geometric split (same gap width, same vertical overlap between
// the two sides). A page with a detected table skips gutter-splitting
// entirely and orders by simple visual position instead, so the table's
// columns are never mistaken for the two sides of a column layout.
export function orderBlocksSimply(blocks: DocumentBlock[]): DocumentBlock[] {
  return sortVisual(blocks.filter((block) => block.text.trim().length > 0));
}
